src/ucs/async/thread.c

```diff
@@ -90,32 +117,13 @@ static void *ucs_async_thread_func(void *arg)
             timeout_ms = ucs_time_to_msec(timer_interval -
                                           ucs_min(time_spent, timer_interval));
         }
-        nready = epoll_wait(thread->epfd, events, UCS_ASYNC_EPOLL_MAX_EVENTS,
-                            timeout_ms);
-        if ((nready < 0) && (errno != EINTR)) {
-            ucs_fatal("epoll_wait() failed: %m");
-        }
-        ucs_trace_async("epoll_wait(epfd=%d, timeout=%d) returned %d",
-                        thread->epfd, timeout_ms, nready);
-
-        /* Check ready files */
-        if (nready > 0) {
-            for (i = 0; i < nready; ++i) {
-                fd = events[i].data.fd;
-
-                /* Check wakeup pipe */
-                if (fd == ucs_async_pipe_rfd(&thread->wakeup)) {
-                    ucs_trace_async("progress thread woken up");
-                    ucs_async_pipe_drain(&thread->wakeup);
-                    continue;
-                }
-
-                status = ucs_async_dispatch_handlers(&fd, 1);
-                if (status == UCS_ERR_NO_PROGRESS) {
-                    is_missed = 1;
-                }
-            }
-        }
+
+        do {
+            status = ucs_event_set_wait(thread->event_set,
+                                        UCS_ASYNC_EPOLL_MAX_EVENTS, timeout_ms,
+                                        ucs_async_thread_ev_handler, cb_arg,
+                                        &read_events);
+        } while ((status == UCS_INPROGRESS) && (read_events != 0));
```

Comment:
why are we doing this in a do..while loop now? there was no loop before..

Reply:
got it. but i think this is a common issue to all users of event_set:
@hiroyuki-sato we need to fix ucs_event_set_wait() to poll for max_events, and not for min(max_events, UCS_EVENT_EPOLL_MAX_EVENTS)
either by using ucs_alloca(max_events*sizeof(int)), or doing the loops inside ucs_event_set_wait

Reply:
yes, this is issues
I was thinking on how we can implement to make it applicable for all scenarios that we have in UCX (TCP and Async). We need to add time measurements (start and begin) for Async needs to wait during given time for a specific number of events (or to handle EINTR error)

Reply:
maybe ucs_event_set_wait() should pass some timeout in 1st iteration, and pass timeout=0 in all next iterations, to read the remaining events

Reply:
yes, this sounds as a good solution here.
so, there is no need to do looping on a USC/event_set user side in this case.
Do we need to read the remaining events until drained?

Reply:
we should read as much as provided by max_events

Reply:
Hello, @yosefe and @dmitrygladkov. Thank you for the review.


I'll work the following steps.
* Keep this PR as work in progress.
* Create new PR for modifying `event_set_wait` on another PR.

Before creating a new PR, I would like to confirm the modification.
I think the modification as discussed @yosefe and @dmitrygladkov like the below, 
However, this modification can't care timeout (The case of no event happen)
What is my misunderstanding?

```C
ucs_status_t ucs_event_set_wait(ucs_sys_event_set_t *event_set,
                                unsigned max_events, int timeout_ms,
                                ucs_event_set_handler_t event_set_handler,
                                void *arg, unsigned *read_events)
{
    struct epoll_event ep_events[UCS_EVENT_EPOLL_MAX_EVENTS];
    unsigned max_wait_events = max_events;
    int nready, events, i;
    int timeout_wait_ms = timeout_ms;

    ucs_assert(event_set_handler != NULL);
    ucs_assert(read_events != NULL);

    *read_events = 0;
    do {
        nready = epoll_wait(event_set->epfd, ep_events, max_wait_events,
                            timeout_wait_ms);

        if (nready < 0 && errno != EINTR ) {
            ucs_error("epoll_wait() failed: %m");
            return UCS_ERR_IO_ERROR;
        }

        ucs_trace_data("epoll_wait(epfd=%d, timeout=%d) returned %d",
                       event_set->epfd, timeout_wait_ms, nready);

        for (i = 0; i < nready; i++) {
            events = ucs_event_set_map_to_events(ep_events[i].events);
            event_set_handler(ep_events[i].data.ptr, events, arg);
        }

        *read_events += nready;
        max_wait_events -= nready;
        /* After 1st iteration, epoll_wait always returns immediately */
        timeout_wait_ms = 0;
    } while ( *read_events < max_events );

    return UCS_OK;
}
```

Reply:
@hiroyuki-sato yes, this is the intention. since the change is not so big, we can do it in this PR, to save some time. 
few minor things to change: 
1. no space line after epoll_wait. 
1. indent +=/-= lines on same column. 
1. add ucs_assert(nready<=max_wait_events) after epoll_wait
4. do { } while loop - `while (max_wait_events > 0)`

Reply:
Hello, @yosefe . Thank you for your comments.

I modified the codes like the following. It this change correct?
If so, I also require to modify test codes.
As I mentioned before, It can't care timeout.
So I need to remove timeout tests (no events happen case).
And I also need to modify the `max_events` size to the number of wait events size explicitly (not MAX_EVENT_SET_SIZE)
https://github.com/hiroyuki-sato/ucx/commit/a0acf38b90ae28f9c0b421609604906d92b963f8#diff-7c9812c959527ed11a9d4ea575c0bc51R98


I pushed this change on another branch now. 
https://github.com/hiroyuki-sato/ucx/commit/a0acf38b90ae28f9c0b421609604906d92b963f8
If it is OK. I commit to this branch. It's OK to create another branch too.


```C
ucs_status_t ucs_event_set_wait(ucs_sys_event_set_t *event_set,
                                unsigned max_events, int timeout_ms,
                                ucs_event_set_handler_t event_set_handler,
                                void *arg, unsigned *read_events)
{
    struct epoll_event ep_events[UCS_EVENT_EPOLL_MAX_EVENTS];
    unsigned max_wait_events = max_events;
    int nready, events, i;
    int timeout_wait_ms = timeout_ms;

    ucs_assert(event_set_handler != NULL);
    ucs_assert(read_events != NULL);

    *read_events = 0;
    do {
        nready = epoll_wait(event_set->epfd, ep_events, max_wait_events,
                            timeout_wait_ms);
        ucs_assert(nready <= max_wait_events);
        if (nready < 0 && errno != EINTR ) {
            ucs_error("epoll_wait() failed: %m");
            return UCS_ERR_IO_ERROR;
        }

        ucs_trace_data("epoll_wait(epfd=%d, timeout=%d) returned %d",
                       event_set->epfd, timeout_wait_ms, nready);

        for (i = 0; i < nready; i++) {
            events = ucs_event_set_map_to_events(ep_events[i].events);
            event_set_handler(ep_events[i].data.ptr, events, arg);
        }

        *read_events    += nready;
        max_wait_events -= nready;
        /* After 1st iteration, epoll_wait always returns immediately */
        timeout_wait_ms = 0;
    } while (max_wait_events > 0);

    return UCS_OK;
}
```

Reply:
@hiroyuki-sato the code above looks good. you can push it with separate PR or in here .. up to you

Reply:
@yosefe Thank you for the comment. I created #3733. It requires to modify test code yet.